crates/dbs-miniball/src/vmm/src/boot.rs

```diff
@@ -0,0 +1,267 @@
+// Copyright 2022 Alibaba Cloud. All Rights Reserved.
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+#![cfg(target_arch = "x86_64")]
+use std::result;
+
+use linux_loader::{bootparam::boot_params, loader::KernelLoaderResult};
+use vm_memory::{Address, GuestAddress, GuestMemory, GuestMemoryMmap};
+
+// x86_64 boot constants. See https://www.kernel.org/doc/Documentation/x86/boot.txt for the full
+// documentation.
+// Header field: `boot_flag`. Must contain 0xaa55. This is the closest thing old Linux kernels
+// have to a magic number.
+const KERNEL_BOOT_FLAG_MAGIC: u16 = 0xaa55;
+// Header field: `header`. Must contain the magic number `HdrS` (0x5372_6448).
+const KERNEL_HDR_MAGIC: u32 = 0x5372_6448;
+// Header field: `type_of_loader`. Unless using a pre-registered bootloader (which we aren't), this
+// field must be set to 0xff.
+const KERNEL_LOADER_OTHER: u8 = 0xff;
+// Header field: `kernel_alignment`. Alignment unit required by a relocatable kernel.
+const KERNEL_MIN_ALIGNMENT_BYTES: u32 = 0x0100_0000;
+
+// Start address for the EBDA (Extended Bios Data Area). Older computers (like the one this VMM
+// emulates) typically use 1 KiB for the EBDA, starting at 0x9fc00.
+// See https://wiki.osdev.org/Memory_Map_(x86) for more information.
+const EBDA_START: u64 = 0x0009_fc00;
+// RAM memory type.
+// TODO: this should be bindgen'ed and exported by linux-loader.
+// See https://github.com/rust-vmm/linux-loader/issues/51
+const E820_RAM: u32 = 1;
+
+#[derive(Debug, PartialEq)]
+/// Errors pertaining to boot parameter setup.
+pub enum Error {
+    /// Invalid E820 configuration.
+    E820Configuration,
+    /// Highmem start address is past the guest memory end.
+    HimemStartPastMemEnd,
+    /// Highmem start address is past the MMIO gap start.
+    HimemStartPastMmioGapStart,
+    /// The MMIO gap end is past the guest memory end.
+    MmioGapPastMemEnd,
+    /// The MMIO gap start is past the gap end.
+    MmioGapStartPastMmioGapEnd,
+}
+
+fn add_e820_entry(
```

Comment:
This function is provided in dbs-boot (or you didn't start to change this part?)

Reply:
I plan to modify it in memory virtualization.

Reply:
Got that